### ISSUE TITLE
Texture memory import API

### ISSUE THREAD

**Issue by zarik5** [gfx-rs/wgpu#2320]
**Is your feature request related to a problem? Please describe.**
In my personal project I need to import texture memory into wgpu (Vulkan backend) from file descriptors on Linux, `HANDLE`s on Windows and `AHardwareBuffer` on Android. fd's and `HANDLE`s are for cross-process interop, `AHardwareBuffer` is for decoder interop.

**Describe the solution you'd like**
I have not designed the API yet.

**Describe alternatives you've considered**
Use the raw handles Vulkan API (initially authored by me). Having a texture memory import API integrated directly into wgpu would remove a lot of boilerplate.

**Comment by i509VCB** [comment]
I'd be willing to help implement this.

My expertise here is on Linux, I don't know how the situation would look on Android or Windows. On Linux the texture memory import/export object is a dmabuf.

I help develop [Smithay](https://github.com/Smithay/smithay), a library for writing Wayland compositors. Smithay heavily relies on texture memory import from dmabufs to implement fast hardware accelerated rendering. I would like to use wgpu as a renderer, but this is one of the big blockers that would make a wgpu renderer something that could be useful in a high performance setting.

Regarding this API, I think it would also make sense to support texture memory export as well.

## API additions

In my opinion the texture memory import/export functionality is generally going to be platform dependent. I don't see any nice way to avoid this. This probably means each platform will need it's own codepath to support these features.

If there are any conventions for platform specific apis, please do mention these.

## Technical details

Dmabuf import/export can be implemented in both EGL and Vulkan given that the requisite extensions are available:

For Vulkan the following extensions are required:
* `VK_EXT_external_memory_dma_buf`
* `VK_EXT_image_drm_format_modifier` (to get a list of valid DRM modifiers)

These Vulkan extensions are supported in the latest NVIDIA beta driver (515 at the time of writing) and generally in Mesa (anv, radv, v3dv (Vulkan 1.1)).

For EGL the following extensions are required:
* `GL_OES_EGL_image` (this would make import/export for an EGLImage possible if anyone would desire such behavior)
* `EGL_EXT_image_dma_buf_import`
* `EGL_EXT_image_dma_buf_import_modifiers`
* `EGL_MESA_image_dma_buf_export`

I believe NVIDIA implements these extensions from what I've heard. (I will edit the message when get a response from someone to run eglinfo for me).

For Linux with Dmabufs I would expose an API something like the following (specifics pending bikeshed and testing the api):

```rust
pub struct PlaneDescriptor {
    pub fd: RawFd,
    pub offset: u32,
    pub stride: u32,
    pub modifier: u64,
}

pub struct DmabufDescriptor {
    plane_0: PlaneDescriptor,
    plane_1: Option<PlaneDescriptor>,
    // up to plane_3 (since a dmabuf has a max of 4 planes)
}

impl Device {
    // This should be able to fail, safety being a concern regarding plane file descriptors.
    // Also could just not be supported by the driver.
    pub unsafe fn import_dmabuf_texture(&self, &DmabufDescriptor) -> Result<Texture, ImportDmabufError> {
        todo!()
    }
}
```

And on `Texture` for export (could be Device if any architectural constraints prevent the code below):

```rust
impl Texture {
    // Similar failure reasons as above, but safe because we create the file descriptors.
    // Also a texture may have a format that cannot be represented as a dmabuf.
    pub fn export_dmabuf(&self) -> Result<Dmabuf, ExportDmabufError> {
        todo!()
    }
}
```

**Comment by i509VCB** [comment]
Texture export is going to be quite interesting to implement:

In Vulkan an image's underlying memory **must** be created with a flag that states what types of memory objects the texture **may** be exported as. (By extending `VkAllocateMemoryInfo` with `VkExportMemoryAllocateInfo`). When exporting an image in vulkan, we export the underlying memory of the texture. Therefore we need to state ahead of time the intention of exporting a texture.

There are a few instances of this requirement in the "Valid Usage" sections of the specification:
- VUID-VkMemoryGetFdInfoKHR-handleType-00671 (Linux Dmabuf export)
- VUID-VkMemoryGetWin32HandleInfoKHR-handleType-00662 (Win32 HANDLE export)
- VUID-VkMemoryGetAndroidHardwareBufferInfoANDROID-handleTypes-01882 (Android AHardwareBuffer export)

To properly support texture export we probably need one of the following:
1. If export is supported, create every texture with the capability to be exported. This has the problem of limiting what formats may be used since Android's format mappings are a subset of the supported Vulkan formats.
2. Allow specifying that a texture may be exported to a platform's memory handle at creation time. (I guess adding a new variant to `TextureFormatFeatureFlags` comes to mind)

I'd prefer the latter for the sake of being explicit.

For EGL with `EGL_MESA_image_dma_buf_export` whether a texture may be exported is entirely implementation dependent (no really the specification states nothing about this (https://www.khronos.org/registry/EGL/extensions/MESA/EGL_MESA_image_dma_buf_export.txt). The best we could do here is probably to *try* to export the image by getting the properties of the texture using `ExportDMABUFImageQueryMESA` before returning the Texture to the user.

I haven't seen any extensions outside of Linux for EGL and exporting texture memory, so EGL would only apply on Linux.

DX12 mentions something about creating the texture's memory in a shared heap, creating a shared handle ([CreateSharedHandle](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-createsharedhandle)) and opening the shared handle ([OpenSharedHandleByName](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-opensharedhandlebyname)) either in another graphics API or process.

DX11 appears to be similar in that case as well.

If wgpu wants to support texture export as a first class part of the API, then we will need to consider the above.

**Comment by zarik5** [comment]
@i509VCB In case you want to export the memory of a texture I think the best route would be to create a separate method for creating the texture with export capabilities, and then export the memory with a method marked as unsafe, and probably exposing the wgpu-hal layer of abstraction first (like my raw Vulkan API) so that some code can be shared if the same functionality is to be implemented in other backends. I think what is important is to make those extension methods distinct from the rest of the safe API because they are not part of the WebGpu standard.

**Comment by mahkoh** [comment]
>VK_EXT_image_drm_format_modifier

This extension is not available for AMD Polaris (GFX8) cards and below which are still widely used: https://gitlab.freedesktop.org/mesa/mesa/-/issues/5882

**Comment by i509VCB** [comment]
Progress update

To summary the work done, things are falling into place for some platforms to support texture memory import, assuming you import the memory for wgpu.

# GL
#3046 exposes what is needed to create external textures for the GL backend on native platforms from GL texture and renderbuffer ids..

### Linux

With this and the previous pull requests to expose internals, there should be enough to import EGLImages into the gles hal and then create a GL texture name and bind the EGLImage to it.

On Linux, this should all work once the pull request is merged. On Linux with Dmabufs, there is no need to use `TEXTURE_EXTERNAL_OES` for the target on at least Mesa (I need info for Nvidia).

On Mesa, even desktop GL supports this use case as `GL_OES_EGL_image` is enabled on Desktop GL by Mesa which could be useful in the future if the gles hal allows selecting GL 4.6 or whatever in the future.

For more info consult this Mesa issue:
- https://gitlab.freedesktop.org/mesa/mesa/-/issues/2119#note_315982

### Android

Android however REQUIRES `TEXTURE_EXTERNAL_OES`. This is a work in progress task that needs to be implemented in Naga and then managed in wgpu-hal.

# Vulkan

Once #3019 is merged, everything should be in place for the Vulkan hal to use externally imported textures. For Vulkan specifically, you will need to manually create the `Instance` and `Device` with the required extensions and the memory import extensions, and then create the `wgpu::Instance` and `wgpu::Device`.

This ***should*** in theory work on every platform that has Vulkan and external memory assuming you import the memory for wgpu.

**Comment by mahkoh** [comment]
TEXTURE_EXTERNAL_OES is required if you want to sample directly from YUV buffers, e.g. from buffers produced by vaapi. Otherwise you have to import every plane of the buffer as a separate texture and write a shader that uses a format-specific algorithm to combine those planes into an RGB image.

**Comment by i509VCB** [comment]
> TEXTURE_EXTERNAL_OES is required if you want to sample directly from YUV buffers, e.g. from buffers produced by vaapi. Otherwise you have to import every plane of the buffer as a separate texture and write a shader that uses a format-specific algorithm to combine those planes into an RGB image.

Well Firefox on Linux is using TEXTURE_2D for YUV and I can't seem to find bugs from that.

**Comment by mahkoh** [comment]
Then firefox is using such a shader either directly or via a library.

>The only exception is when importing directly as YUV. In that case, many users (e.g. Weston) still avoid using TEXTURE_EXTERNAL: instead of importing as YUV, **they import the luma & chroma planes separately as R/RG/ARGB images bound to TEXTURE_2D and convert in a fragment shader.**

From the very mesa issue you linked.

**Comment by i509VCB** [comment]
Ah, so YUV needs the special sampler then. I checked and firefox does indeed decompose the images into parts.

**Comment by jgcodes2020** [comment]
Came up on this issue; is it still planned or is it currently waiting on someone to implement?

**Comment by expenses** [comment]
I've been looking into importing DMA buffers into `wgpu_hal::vulkan::Texture`s  as a first step towards this and have written this function: https://github.com/gfx-rs/wgpu/compare/wgpu-hal-v26.0.4...expenses:wgpu:texture_from_dma. It runs without validation errors, but for my particular project I'm just getting a white texture so it's not very conclusive 😅 . It'd be great to have someone who knows more about DMA to look over this.

`eglExportDMABUFImageQueryMESA` returns a modifier value of `72057594037927935`/`0xFFFFFFFFFFFFFF` for me which is invalid according to the `drm-fourcc` crate: https://github.com/dzfranklin/drm-fourcc-rs/blob/1ad1a978a9aaa78ff0016cc7013787e4d1d68682/src/consts.rs#L163. I'm not sure at all how to handle this, the documentation gets pretty sparse here.

**Comment by i509VCB** [comment]
> I've been looking into importing DMA buffers into `wgpu_hal::vulkan::Texture`s as a first step towards this and have written this function: [wgpu-hal-v26.0.4...expenses:wgpu:texture_from_dma](https://github.com/gfx-rs/wgpu/compare/wgpu-hal-v26.0.4...expenses:wgpu:texture_from_dma).

I would probably use OwnedFd (Although it will require a cfg guard on windows) to specify ownership. I recall that Vulkan takes ownership of fds which you import.

The API is probably fine for normal image formats. But where it gets annoying fast is YCbCr formats and some tilings where you need to import multiple planes. Don't need to address it immediately, but give it a little thought. 

> I'm just getting a white texture so it's not very conclusive 😅 .

If you are sharing menory, especially across processes you need to make sure that the proper synchronization barriers are used in Vulkan when the imported texture is used. Usually this involves a barrier telling the driver with an external or foreign queue family transfer that it needs to do something to get the image data. 

> `eglExportDMABUFImageQueryMESA` returns a modifier value of `72057594037927935`/`0xFFFFFFFFFFFFFF` for me 

I'm not sure if invalid is returned because you are using an emulated format (and therefore isn't exportable)? Maybe see if other texture formats give you something you can work with?

**Comment by mahkoh** [comment]
>I've been looking into importing DMA buffers into wgpu_hal::vulkan::Textures as a first step towards this and have written this function: https://github.com/gfx-rs/wgpu/compare/wgpu-hal-v26.0.4...expenses:wgpu:texture_from_dma. It runs without validation errors, but for my particular project I'm just getting a white texture so it's not very conclusive 😅 . It'd be great to have someone who knows more about DMA to look over this.

There are several errors in there. You can take a look at this function to see how to properly import dmabufs into vulkan: https://github.com/mahkoh/jay/blob/636b24234750b659f66d0c7d619ec8e6118cf90d/src/gfx_apis/vulkan/image.rs#L284

Some notes:

- self.disjoint is true if and only if the plane file descriptors don't all refer to the same file description
- You can assume that self.render_needs_bridge is false for simplicity

>eglExportDMABUFImageQueryMESA returns a modifier value of 72057594037927935/0xFFFFFFFFFFFFFF for me which is invalid according to the drm-fourcc crate: [dzfranklin/drm-fourcc-rs@1ad1a97/src/consts.rs#L163](https://github.com/dzfranklin/drm-fourcc-rs/blob/1ad1a978a9aaa78ff0016cc7013787e4d1d68682/src/consts.rs#L163). I'm not sure at all how to handle this, the documentation gets pretty sparse here.

You don't. The invalid modifier is not supported in vulkan. If you are on AMD polaris or earlier, you cannot use vulkan dmabuf import/export.

>The API is probably fine for normal image formats. But where it gets annoying fast is YCbCr formats and some tilings where you need to import multiple planes. Don't need to address it immediately, but give it a little thought.

On AMD almost all images use at least 2 planes in my experience.

**Comment by expenses** [comment]
@i509VCB @mahkoh thanks for your suggestions! I'll look into those links later.

I wrote that code with the intention of using it for importing OpenGL textures, but I've since realized that I can use share a Vulkan texture with OpenGL and do a blit inside OpenGL to achieve the same thing but in a much less fiddly way. https://bxt.rs/blog/fast-half-life-video-recording-with-vulkan/ is the best source I've found on how to do this.

https://github.com/gfx-rs/wgpu/compare/wgpu-hal-v26.0.4...expenses:wgpu:create_shareable_texture is the branch I've created for that. It's relatively simple, though a better API might involve passing a `(file_descriptor, allocation_size, semaphore)` closure to the function or something similar.

**Comment by mahkoh** [comment]
I wouldn't use opaque fds. They cannot be imported back into vulkan unless the driver version, driver UUID, and image creation parameters are exactly the same. Otherwise the behavior is undefined. And import is across drivers is going to fail in practice because mesa hardware drivers, llvmpipe, and nvidia (AIUI) all use different file types for opaque fds.

dmabufs are the only universally supported transport on linux.

**Comment by i509VCB** [comment]
create_shareable_texture would involve wgpu needing to step into the slightly complicated process of creating textures which are truly shareable. I am unsure if wgpu is the right place for exporting images, but certainly should support import.

If you are curious about export on Linux, Smithay's Vulkan allocator might be a good reference (for import and export) https://github.com/Smithay/smithay/blob/master/src/backend/allocator/vulkan/mod.rs

On Linux with dmabuf exportable images which you plan to do something with like give to gstreamer or display hardware (especially the case on M1/M2 asahi linux where the GPU and DCP are two different devices to the kernel) need to do proper modifier and format negotiation. Usually this means you ask both ends what format + modifier combinations are supported and pick the best format in a union of that (Vulkan will pick the most optimal combo you provide in a list). wgpu without effectively copying the Vulkan allocator from smithay will be unable to negotiate format and modifier combinations properly.

I would probably not prioritize opaque FDs unless you have a particular use case for it (some embedded Linux setup which doesn't have dmabuf, but does have opaque fds)

**Comment by expenses** [comment]
Glad to be having this discussion. I'm only exporting the opaque FD for the purpose of writing to the texture in OpenGL, though I suppose that distinction doesn't matter.

Is there a crate for getting the dma buffer FD etc. from an OpenGL texture? It's a mesa/egl thing so I'm not sure if gl_generator etc generates it. I'm using raw bindings and might be doing something wrong there which is why I'm getting the invalid modifier.

**Comment by mahkoh** [comment]
https://registry.khronos.org/EGL/extensions/MESA/EGL_MESA_image_dma_buf_export.txt

**Comment by alltheseas** [comment]
I'm trying to build a zero-copy video player on egui - https://github.com/egui-vid/egui-vid/issues/8.
                                                                                                                                                                                          
Video player to use VK_ANDROID_external_memory_android_hardware_buffer. 

After investigating wgpu-hal, found two architectural blockers:            
                                                                                                                                                                                          
  1. Immutable Samplers (for YCbCr conversion)                                                                                                                                            
                                                                                                                                                                                          
  VK_KHR_sampler_ycbcr_conversion requires immutable samplers in VkDescriptorSetLayoutBinding. Currently wgpu-hal always sets p_immutable_samplers: ptr::null() (device.rs:1311).         
                                                                                                                                                                                          
  Question: Is there appetite for adding immutable sampler support to wgpu's bind-group model? Or would a different approach (e.g., baking YCbCr conversion into a separate blit pass) be preferred?                                                                                                                                                                              
                                                                                                                                                                                          
  2. VK_FORMAT_UNDEFINED for external formats                                                                                                                                             
                                                                                                                                                                                          
  External AHardwareBuffers use VK_FORMAT_UNDEFINED with VkExternalFormatANDROID. wgpu's format mapping (conv.rs) only handles concrete formats.                                          
                                                                                                                                                                                          
  Question: Would a new internal format variant (e.g., TextureFormat::ExternalAndroid) or special-case validation path be acceptable?                                                     
                                                                                                                                                                                          
  Reference                                                                                                                                                                               
                                                                                                                                                                                          
  - Existing win32 external memory pattern in device.rs:540-600                                                                                                                           
  - Jay compositor and Smithay allocator implementations                                                                                                                                  
  - Features::TEXTURE_FORMAT_NV12 and ExternalTextureDescriptor already exist

**Comment by HigherOrderLogic** [comment]
I'm planning to create a PR to add a function equivalent to `Device::texture_from_dmabuf_fd` but for OpenGl. Would that be accepted or there are other plans?
